keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,537 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a way to set different RBACs to `finalizers` field and its entire resource.
+
+## Motivation
+
+`Finalizers` is used by controllers as a pre-deletion hook.
+Currently, `finalizers` field isn't subresourced, therefore RBAC for the field can only be set through the one for its entire resource.
+On the other hand, some controllers only need to update `finalizers` field and some of its subresources, and don't need to update its entire resource.
+To implement least privilege for such controllers, there should be a way to only allow updating `finalizers`.
+
+### Goals
+
+- Provide a way to set a different RBAC to `finalizers` from its entire resource
+
+### Non-Goals
+
+- Provide a way to set a different RBAC to other than `finalizers` from its entire resource
+- Provide a way to allow updating its entire resource  but disallow updating `finalizers` field
+
+## Proposal
+
+A new endpoint `finalize` is added as an alternative way to access to `finalizers` field and allow setting a separate RBAC to the endpoint.
+The existing way to acceess `finalizers` field through `metadata.finalizers` remains unchanged.
+
+### User Stories (Optional)
+
+#### Story 1
+
+There is a controller that only updates `status` field and `finalizers` field of a CRD. Administrators of the controller can allow the controller to update only these particular fields.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Name collision of the endpoints is needed to be considered.
+- Other existing fields may require the same alternative way to access.
+For example, [candidates](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go) in metadata fields are:
+  - `labels`
+  - `annotations`
+  - `ownerReferences`
+- Depending on the implementation, read access to the entire resource, not just its subresource, may be required to patch the subresource. Because patch operation requires read access to ensure that it doesn't change any other part of the resource.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
```

Comment:
This needs to detail the schema of the payload and the verbs the subresource allows.

E.g. can I POST to `/{resource_name}/finalize` or just PUT?  Does PATCH work?  What happens if I try to DELETE?

If I GET it, what kind of object do I get back (and is that the same as when I PUT)?

We can look at what /scale does: it allows GET, PATCH, PUT

Reply:
for example, it could simply be a list of strings which are added to the `finalizers` with set semantics (if it existed already, not an error).

Also need to spec how I remove an item.

e.g.

```
{
    "add": [ "example.com/my-finalizer" ],
    "del": [ "example.com/old-finalizer" ]
}
```

Reply:
Thank you for sharing your ideas on API definition. My prototype was too much hacky to achieve it with minimum code changes. I agree that defining a new proper API to handle this use case will be better. So, I will update this KEP based on your idea.

(As explained in "Notes/Constraints/Caveats (Optional)" section, Get call in my prototype is broken (it returns entire resource, not just subresource, even if GET for entire resource isn't allowed). It will be fixed if we defined a proper API).

Reply:
It looks like you're just using "anything that can satisifies `Object` as your schema, but I think we can be more precise (and removing items is a critical use case!!)

Reply:
+1 on deletion - I would even say, it's "the most critical usecase").

I like the "add", "del" fields (maybe we can figure out better names).
What I was wondering though is if would make any sense to try making them generic (so that the further fields, like labels and annotations could follow the same). But maybe if we limit the commonality to the structure of "add/delete" fields and let each fields to define its own type for those is actually enough.

Reply:
I was just handwaving.  I wrote the below comment, but I think I like Wojciech's idea better.  I will leave this for history :)

Maybe every subresource should define its own schema, but use the same patterns.  All the items listed so far (liens, labels, annotations, ownerRefs) are n-ary.  Do we think we'll ever see a scalar value ?  I'd bet "yes" in the fullness of time.

I am not an expert here, and PATCH in particular has a ton of baggage

So something like:

GET .../foo/liens
* request: {}
* response: { value: []string }

PUT .../foo/liens
* request: { value: []string }
* response: { value: []string }

PATCH .../foo/liens
* request: { add: []string, delete: []string }
* response: { value: []string }

but labels would be:

GET .../foo/labels
* request: {}
* response: { value: map[string]string }

PUT .../foo/labels
* request: { value: map[string]string }
* response: { value: map[string]string }

PATCH .../foo/labels
* request: { add: map[string]string, delete: []string }
* response: { value: mape[string]string }

and some hypothetical `int32` scalar value would be a little different

GET .../foo/width
* request: {}
* response: { value: int32 }

PUT .../foo/width
* request: { value: int32 }
* response: { value: int32 }

PATCH .../foo/width
* request: { value: int32 }
* response: { value: int32 }

Do we even need to support PUT?  Or should PUT *and* PATCH both be "add"/"del"?

Also, it looks like /scale and others carry objectmeta, so we should too.

Scale also has "status" which I don't think we need here.  See `func scaleFromRC` for example

Reply:
so, as a "sub**resource**", shouldn't this behave as a **resource**. So what is listed under `finalizers` are treated as actual named resources (that happen to not be *real* objects). Something like:

`GET /{resource_name}/finalizers` - returns the collection of finalizers for `resource_name`

`PUT /{resource_name}/finalizers` - writes the collection of finalizers for `resource_name`

`PUT /{resource_name}/finalizers/example.com/my-finalizer` - adds `example.com/my-finalizer` to the collection of finalizers for `resource_name`

`DELETE /{resource_name}/finalizers` - deletes the **collection** of finalizers for `resource_name`

`DELETE /{resource_name}/finalizers/example.com/my-finalizer` - deletes the  `example.com/my-finalizer` from the collection of finalizers for `resource_name`

`PATCH` would only apply to the collection `.../finalizers`, I would think, and would effectively be a multi-value `PUT` that doesn't clobber existing values.

There would be no multi-value `DELETE`, those need to be done one-by-one or entire collection.

Reply:
I'm fairly opposed to inventing new semantics when we already have all the mechanisms to do this things?
I see people doing:
1. RMW (read-modify-write) using resource-version, optimistic locking, we all know how this works,
2. Patches in general are not great since they don't deal with lists very well, neither json patch nor json-merge-patch are great. json-merge-patch forces a full-replace of the list (so basically equivalent to RMW), and json patch can delete by index, but then you need to know the exact ordering of the list, which goes back to RMW.
3. Server-side apply works great for that use-case, but only if you have applied the finalizer in the first place (and dealing with managedfields in the sub-resource can be tedious, though we've done it for scale, btw, that should probably be part of this proposal).

`labels` are entirely different since they are a map, items can be deleted individually either through json merge patch or json-patch (which already defines the "delete" operation to remove a specific item). So what you're proposing already exists.

For the type of the returned object `GET /{resource_name}/finalizers`, possibly we could return the `ObjectMeta` (since it already exists), contains the object identifier, and the finalizers. Only changes to the finalizers would be honored (similar to what we do for `/scale`)

EDIT after reading below: I see you're discussing the type, thanks.

Reply:
I expect this subresource to be exactly metadata, with only finalizers field mutable, and you can use SSA or any other technique to set it. Hopefully that's what I'll find below.

Reply:
> Hopefully that's what I'll find below.

Narrator: He did not.

Reply:
> I expect this subresource to be exactly metadata, with only finalizers field mutable, and you can use SSA or any other technique to set it. Hopefully that's what I'll find below.

That's roughly what I suggested in https://github.com/kubernetes/enhancements/pull/3286#discussion_r899778508 (it's not the case in the proposal now).